src/engine/EntityComponentSystem/Components/TransformComponent.ts

```diff
@@ -20,15 +74,110 @@ export enum CoordPlane {
 export class TransformComponent extends Component<'transform'> {
   public readonly type = 'transform';
 
+  private _dirty = false;
+  // private _globalDirty =  false;
+
+  public readonly matrix = Matrix.identity()
+    .translate(0, 0)
+    .rotate(0)
+    .scale(1, 1)
+  private _position = createPosView(this.matrix);
+  private _rotation = 0;
+  private _scale = createScaleView(this.matrix);
+
+
+  private _recalculate() {
+    // this._position = new VectorMatrixPosView(this._mat);
+    this._rotation = this.matrix.getRotation();
+    // this._scale = new VectorMatrixScaleView(this._mat);
+    this._dirty = false;
+  }
+
+  // private _globalMat: Matrix;
+  // private _recalculateGlobal() {
+  //   this._globalMat = this.getGlobalMatrix();
+  //   this._globalDirty = false;
+  // }
+
+  public getGlobalMatrix(): Matrix {
+    if (!this.parent) {
+      return this.matrix;
+    } else {
+      return this.parent.getGlobalMatrix().multm(this.matrix);
+    }
+  }
+
+  public get parent(): TransformComponent | null {
+    return this?.owner?.parent?.get(TransformComponent);
+  }
+
   /**
    * The [[CoordPlane|coordinate plane|]] for this transform for the entity.
    */
   public coordPlane = CoordPlane.World;
 
+
   /**
-   * The current position of the entity in world space or in screen space depending on the the [[CoordPlane|coordinate plane]]
+   * The current position of the entity in world space or in screen space depending on the the [[coordinate plan|CoordPlane]].
+   *
+   * If a parent entity exists coordinates are local to the parent.
    */
-  public pos: Vector = Vector.Zero;
+  public get pos(): Vector {
+    if (this._dirty) {
+      this._recalculate();
+    }
+    return this._position;
+  }
+
+  public set pos(val: Vector) {
+    this.matrix.setPosition(val.x, val.y);
+    this._dirty = true;
+  }
+
+  // Dirty flag check up the chain
+  public get dirty(): boolean {
+    if (this?.owner?.parent) {
+      const parent = this.parent;
+      return parent.dirty || this._dirty;
+    }
+    return this._dirty;
+  }
+
+  /**
+   * The current world position calculated
+   */
+  public get globalPos(): Vector {
+    return new VectorView({
+      data: this.getGlobalMatrix(),
+      getX: (source) => source.data[12],
+      getY: (source) => source.data[13],
+      setX: (source, x) => {
+        if (this.parent) {
+          const [ newX ] = this.parent?.getGlobalMatrix().getAffineInverse().multv([x, source.data[13]]);
+          this.matrix.data[12] = newX;
+        } else {
+          this.matrix.data[12] = x;
+        }
+      },
+      setY: (source, y) => {
+        if (this.parent) {
+          const [, newY ] = this.parent?.getGlobalMatrix().getAffineInverse().multv([source.data[12], y]);
+          this.matrix.data[13] = newY;
+        } else {
+          this.matrix.data[13] = y;
+        }
+      }
+    });
+  }
```

Comment:
Yeah, with these numbers appearing all over, they should be made into constants.

Reply:
Would it make sense to add getters for whatever this data array/value is? Should something outside the matrix be mutating it? It might be safer in the long run to get/set so consuming code doesn't have to know where `12` or `13` are in the data.